### PR TITLE
test: Testing stack + `Hit` and `Template` tests

### PR DESCRIPTION
**Pull request by vvo**
The testing stack is ready.

It's composed of:
- karma
- mocha
- expect (assertions)

Changes in npm scripts:
- `test` now runs the tests in phantomjs and then lint
- `test:watch` will open a real chrome canary, output results in terminal console and auto reload tests

It's basically the stack from https://github.com/rackt/react-router/ with some modifications.

I wrote the `Hits` and `Template` tests, they may not be exhaustive but that's a start. Once we get this merged I will create the relevant issues so that anyone can help in writing tests.

We need to strengthen this library because untested code will and must #fail.

TODO:
- [ ] port test/ folder to karma, mocha, expect

### CODE REVIEW

**Comment by vvo** [comment]
Currently rewriting the previous tests to follow the same pattern

File: components/Hits.js [hunk 1]
```diff
@@ -1,11 +1,9 @@
 var React = require('react');
 var map = require('lodash/collection/map');
 
-var Template = require('./Template');
-
 class Hits extends React.Component {
   renderWithResults() {
-    var renderedHits = map(this.props.hits, (hit) => {
+    var renderedHits = map(this.props.results.hits, (hit) => {
```
Comment (bobylito) [inline components/Hits.js]: Why choose to have a results instead of just hits?

File: components/Template.js [hunk 2]
```diff
@@ -69,4 +67,44 @@ function transformData(fn, templateKey, originalData) {
...
+
+    content = hogan.compile(template, compileOptions).render(data);
+  }
+
+  // We add all our template helper methods to the template as lambdas. Note
+  // that lambdas in Mustache are supposed to accept a second argument of
+  // `render` to get the rendered value, not the literal `{{value}}`. But
+  // this is currently broken (see
+  // https://github.com/twitter/hogan.js/issues/222).
+  function addTemplateHelpersToData(templateData) {
```
Comment (bobylito) [inline components/Template.js]: I would extract this function to the outer scope to make the return more obvious.

**Comment by redox** [comment]
Why did you choose `components/__tests__/mycomponent.js` over `test/components/mycomponent.js`?

File: components/Template.js [hunk 3]
```diff
@@ -69,4 +67,44 @@ function transformData(fn, templateKey, originalData) {
   throw new Error('`transformData` must be a function or an object');
 }
 
+function renderTemplate({template, compileOptions, helpers, data}) {
+  var hogan = require('hogan.js');
+  var forOwn = require('lodash/object/forOwn');
+  var content;
+
+  if (typeof template !== 'string' && typeof template !== 'function') {
```
Comment (bobylito) [inline components/Template.js]: Even though `typeof` is pretty common, it's still not very readable. What about : 
```js
var isString = typeof template === 'string';
var isFunction = typeof template === 'function';

if(!isString && !isFunction) {
  throw new Error('Template must be `string` or `function`');
}
content = isFunction ? 
  template(data):
  hogan.compile(template, compileOptions).render(addTemplateHelpersToData(data));

```

File: components/Hits.js (hunk 1 above)
Comment (vvo) [inline components/Hits.js]: While writing the tests I wondered why we where splitting the results object in two (hits: results.hits, results: results). If we already are dealing with the whole `results` then we might just pass it and use it inside the component. Since it already knows about the whole `results` object.

File: components/Template.js (hunk 2 above)
Comment (vvo) [inline components/Template.js]: Ah I should have said that, I did not modify this function, we can refactor it later on but I did not touch it. I only ported it from utils/ to Template, since it was the only place using it.

File: components/Template.js (hunk 3 above)
Comment (vvo) [inline components/Template.js]: same as the other refactor component

**Comment by vvo** [comment]
> Why did you choose components/__tests__/mycomponent.js over test/components/mycomponent.js?

So this is something I borrowed from the various famous react projects I looked at before doing our tests. Like react itself (https://github.com/facebook/react/tree/master/src/renderers/dom), react-router (https://github.com/rackt/react-router/tree/master/modules) and most rackt organization projects (https://github.com/rackt).

I found it handy and good because:
- tests are close to the implementation, in term of file organization
- you can cleary see if they are missing tests because if there's one missing test file then it's easier because there's less hierarchy to dig into
- we follow a convention that seems to get some traction in most react or not react based projects

It also look like what golang is doing (file.go, file_test.go). Maybe we could directly take this road which is even better IMHO.

Good for you?

**Comment by redox** [comment]
> It also look like what golang community is doing (file.go, file_test.go). Maybe we could directly take this road which is even better IMHO.

I generally don't put them too close because otherwise it's not clear where to put fixture/resource files that are used by tests (you clearly don't want to add them next to sources). Having the `__tests__` directory is OK if that's the "default/common" convention for React, the naming is just super weird -> it really looks like something auto-generated or that you shouldn't modify :/

**Comment by vvo** [comment]
> Having the __tests__ directory is OK if that's the "default/common" convention for React, the naming is just super weird -> it really looks like something auto-generated or that you shouldn't modify :/

Yes at first look it seems weird I agree.

**Comment by vvo** [comment]
> I generally don't put them too close because otherwise it's not clear where to put fixture/resource files that are used by tests (you clearly don't want to add them next to sources).

I guess the golang community is no big fan of heavy fixtures or just put them inside the test file. If you know how to collapse code blocks with keyboard thats ok :D (ctrl+k[+0/1/2] in st)

**Comment by vvo** [comment]
Let's use __tests__ for now

File: components/Template.js (hunk 2 above)
Comment (bobylito) [inline components/Template.js]: Erf, wrong move then :P Ok let's do that later then.

**Comment by redox** [comment]
LGTM :+1:

**Comment by vvo** [comment]
Pr was not done lol I thought it was clear 😀

**Comment by vvo** [comment]
Will revert and reopen later

**Comment by redox** [comment]
Arg :/ My bad -> maybe easier to open a new one for the `test/` folder port

**Comment by vvo** [comment]
no prob

**Comment by vvo** [comment]
Will add a "do-not-merge" label

**Comment by vvo** [comment]
2015-10-10-224208_273x72_scrot

:D

**Comment by bobylito** [comment]
`on hold` is pretty common ;)

File: components/Template.js (hunk 3 above)
Comment (pixelastic) [inline components/Template.js]: Agreed with both of you: `isString`/`isFunction` is much more readable, but this should be amended in another commit.
